File: julia/5_mps2statevector.jl
#!/usr/bin/env julia
# -*- coding: utf-8 -*-
"""
Generate statevector from MPS (Julia)
"""

using TensorOperations

# Python の reshape(-1)（行優先）に合わせて 2D を一次元化する補助関数
cflatten(M::AbstractMatrix) = vec(permutedims(M))  # 行→列入替後に vec

function main()
    # Bell state
    println("Bell state:")
    tl = [1.0 0.0; 0.0 1.0] / 2^(0.25)
    tr = [1.0 0.0; 0.0 1.0] / 2^(0.25)
    println("left tensor: ", tl)
    println("right tensor: ", tr, "\n")

    @tensor bell[i, k] := tl[i, j] * tr[j, k]
    # Python の reshape(-1) と同順にしたい場合は cflatten を使う
    state_bell = cflatten(bell)
    println("statevector: ", state_bell, "\n")

    # GHZ state
    n = 6
    println("n=$(n) GHZ state:")
    w = 2^(1 / (2n))
    tl = [1.0 0.0; 0.0 1.0] / w
    tr = [1.0 0.0; 0.0 1.0] / w
    t = zeros(Float64, 2, 2, 2)  # (j,k,l)
    t[1, 1, 1] = 1 / w
    t[2, 2, 2] = 1 / w
    println("left tensor: ", tl)
    println("right tensor: ", tr)
    println("middle tensors: ", t, "\n")

    # 連結（縮約）: A(ij) と T(jkl) -> (i k l) にしてから (i*k, l) に reshape
    ghz = tl
    for _ = 2:(n-1)  # Python の range(1, n-1) と同じ回数 (n-2回)
        @tensor tmp[i, k, l] := ghz[i, j] * t[j, k, l]
        ghz = reshape(tmp, size(tmp, 1) * size(tmp, 2), size(tmp, 3)) # (i*k, l)
    end
    @tensor ghz2[i, k] := ghz[i, j] * tr[j, k]
    state_ghz = cflatten(ghz2)
    println("statevector: ", state_ghz, "\n")
end

main()

File: rust/src/bin/5_mps2statevector.rs
use ndarray::{Array2, Array3, ArrayD, array};
use ndarray_einsum::einsum;
use tn_basics::MapStrToAnyhowErr;

// Generate statevector from MPS

#[expect(clippy::cast_precision_loss)]
fn main() -> anyhow::Result<()> {
    // Bell state
    println!("Bell state:");
    let tl: Array2<f64> = array![[1.0, 0.0], [0.0, 1.0]] / 2f64.powf(0.25);
    let tr: Array2<f64> = array![[1.0, 0.0], [0.0, 1.0]] / 2f64.powf(0.25);
    println!("left tensor:\n{tl}");
    println!("right tensor:\n{tr}\n");

    let bell = einsum("ij,jk->ik", &[&tl, &tr]).map_str_err()?;
    let state_bell = bell.flatten();
    println!("statevector:\n{state_bell}\n");

    // GHZ state
    let n: usize = 6;
    println!("n={n} GHZ state:");
    let w = 2f64.powf(1.0 / (2.0 * n as f64));
    let tl: Array2<f64> = array![[1.0, 0.0], [0.0, 1.0]] / w;
    let tr: Array2<f64> = array![[1.0, 0.0], [0.0, 1.0]] / w;
    let mut t = Array3::zeros((2, 2, 2));
    t[[0, 0, 0]] = 1.0 / w;
    t[[1, 1, 1]] = 1.0 / w;
    println!("left tensor:\n{tl}");
    println!("right tensor:\n{tr}");
    println!("middle tensors:\n{t}\n");

    let mut ghz: Array2<f64> = tl;
    for _ in 1..(n - 1) {
        let tmp: ArrayD<f64> = einsum("ij,jkl->ikl", &[&ghz, &t]).map_str_err()?;
        ghz = tmp
            .to_shape((tmp.shape()[0] * tmp.shape()[1], tmp.shape()[2]))?
            .into_owned();
    }
    let ghz = einsum("ij,jk->ik", &[&ghz, &tr]).map_str_err()?;
    let state_ghz = ghz.flatten();
    println!("statevector:\n{state_ghz}\n");

    Ok(())
}
